pam/pam_test/module-transaction-dummy_test.go

```diff
@@ -0,0 +1,512 @@
+					func(style pam.Style, msg string) (string, error) {
+						convFunCalled = true
+						require.Equal(t, prompt, msg)
+						require.Equal(t, tc.convStyle, style)
+						return tc.expected, tc.convError
+					})
+			}())
+
+			var reply pam.StringConvResponse
+			var err error
+
+			if tc.promptFormat != "" {
+				reply, err = tx.StartStringConvf(tc.convStyle, tc.promptFormat,
+					tc.promptFormatArgs...)
+			} else {
+				reply, err = tx.StartStringConv(tc.convStyle, tc.prompt)
+			}
+
+			require.ErrorIs(t, err, tc.expectedError)
+			expectConFuncCalled := !tc.convShouldNotBeCalled && tc.convHandler == nil
+			require.Equal(t, expectConFuncCalled, convFunCalled)
+
+			if tc.expectedError == nil {
+				require.NotNil(t, reply)
+				require.Equal(t, tc.expected, reply.Response())
+				require.Equal(t, tc.convStyle, reply.Style())
+			} else {
+				require.Zero(t, reply)
+			}
+		})
+	}
+}
+
+func TestStartBinaryConv(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(MaybeDoLeakCheck)
+
+	testCases := map[string]struct {
+		request       []byte
+		convError     error
+		convHandler   *pam.BinaryConversationHandler
+		expected      []byte
+		expectedError error
+	}{
+		"simple": {
+			request:  []byte{0x01, 0x02, 0x03},
+			expected: []byte{0x00, 0x01, 0x02, 0x03, 0x4},
+		},
+
+		"with error": {
+			request:       []byte{0x03, 0x02, 0x01},
+			convError:     pam.BufErr,
+			expectedError: pam.BufErr,
+		},
+
+		"no conv set fails": {
+			convHandler:   ptrValue(pam.BinaryConversationHandler(nil)),
+			expectedError: pam.ConvErr,
+		},
+	}
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Cleanup(MaybeDoLeakCheck)
+
+			convFunCalled := false
+			var tx pam.ModuleTransaction
+			if tc.convHandler != nil {
+				tx = NewModuleTransactionDummy(*tc.convHandler)
+			} else {
+				tx = NewModuleTransactionDummy(func() pam.BinaryConversationFunc {
+					return pam.BinaryConversationFunc(
+						func(ptr pam.BinaryPointer) ([]byte, error) {
+							convFunCalled = true
+							require.NotNil(t, ptr)
+							bytes := getBinaryFromPointer(ptr, len(tc.request))
+							require.Equal(t, tc.request, bytes)
+							return tc.expected, tc.convError
+						})
+				}())
+			}
+			response, err := tx.StartBinaryConv(tc.request)
+			require.ErrorIs(t, err, tc.expectedError)
+			require.Equal(t, tc.convHandler == nil, convFunCalled)
+
+			if tc.expectedError == nil {
+				defer response.Release()
+				require.NotNil(t, response)
+				require.Equal(t, pam.BinaryPrompt, response.Style())
+				require.NotNil(t, response.Data())
+				bytes, err := response.Decode(func(ptr pam.BinaryPointer) ([]byte, error) {
+					return *(*[]byte)(ptr), nil
+				})
+				require.NoError(t, err)
+				require.Equal(t, tc.expected, bytes)
+
+				bytes, err = response.Decode(nil)
+				require.ErrorContains(t, err, "nil decoder provided")
+				require.Nil(t, bytes)
+			} else {
+				require.Nil(t, response)
+			}
```

Comment:
I know this feels like "preference", but it's more go-like to keep the "happy-path" and have your code as aligned to the left as possible.

```suggestion
			if tc.expectedError != nil {
				require.Nil(t, response)
				return
			}
			defer response.Release()
			require.NotNil(t, response)
			require.Equal(t, pam.BinaryPrompt, response.Style())
			require.NotNil(t, response.Data())
			bytes, err := response.Decode(func(ptr pam.BinaryPointer) ([]byte, error) {
				return *(*[]byte)(ptr), nil
			})
			require.NoError(t, err)
			require.Equal(t, tc.expected, bytes)

			bytes, err = response.Decode(nil)
			require.ErrorContains(t, err, "nil decoder provided")
			require.Nil(t, bytes)
```